Store/retrieve sample data from SQL database

jorainer opened the pull request:
This PR adds the functionality to store and retrieve sample data (and sample to spectra mapping) to and from a `MsBackendSql` database. See also https://github.com/rformassspectrometry/MsBackendSql/issues/14 for the initial idea and discussion.

In this PR:

- add `dbWriteSampleData` to store sample data from a `MsExperiment` to the SQL database of the `Spectra` object of the `MsExperiment`. This requires that the `Spectra` object uses a `MsBackendSql` or `MsBackendOfflineSql`.
- modify the `MsExperiment` constructor function to take also values for the individual slots. If a `Spectra` with backend inheriting `MsBackendSql` is used, and the respective data is available in the database, sample data and mapping samples to spectra is retrieved from the database.

This facilitates and simplifies creation of self-contained, distributable and language-agnostic representations of an `MsExperiment` (MS data and sample data are stored within a SQL database).

This closes issue #39

jorainer committed:
feat: add functionality to get sample data from SQL

- Add functionality to store/read sample data to/from a MsBackendSql
  database (issue #39).

jorainer committed:
docs: fix documentation

jorainer committed:
docs: update documentation and add section to vignette

- Update/fix documentation.
- Add a section to the vignette describing use of `MsExperiment` with
  `MsBackendSql`.

jorainer committed:
ci: use R devel for macOS and windows

andreavicini reviewed (approved):
Everything looks great to me :)

jorainer commented:
Thanks @andreavicini - are you also OK with naming of the functions? Maybe you find a better nomenclature? Also, I will need to fix now the failing unit tests.

andreavicini commented:
> Thanks @andreavicini - are you also OK with naming of the functions? Maybe you find a better nomenclature? Also, I will need to fix now the failing unit tests.

Oh really, failing unit tests?! For the functions I guess you used the same naming style as the functions from DBI. It’s OK for me :)

philouail commented on R/MsExperiment-db.R:
```diff
@@ -0,0 +1,166 @@
+            unname(as.matrix(
+                dbGetQuery(con, "select * from sample_to_msms_spectrum")))
+        else matrix(ncol = 2, nrow = 0)
+    } else matrix(ncol = 2, nrow = 0)
```
the structure of this "link matrix" is 2 column, one for the sample ID and one for the spectra IDs  ?

philouail replied:
actually I am a bit confused with the annotation "sample_to_msms_spectrum". is it only for MS2 ? If it is, then I have a hard time understanding the code could you explain what the different steps do here

jorainer replied:
yes, the link matrix is a two column matrix with the indices (IDs/primary keys) to link samples and spectra - in the form:


| sample | spectra |
|--|--|
| 1 | 1 |
| 1 | 2 |
| 1 | 3 |
| 1 | 4 |
| 2 | 5 |
| 2 | 6 |
| ... | ... |

this allows us also to have n:m relationships (although at present we use only 1:n, i.e. one sample has many spectra, but each spectra is associated only to one sample).

And yes, agree, the name might be a bit misleading. In fact, the database has tables called *sample_data* (for the sample annotations), *msms_spectrum* (for the spectra variables, one row for each spectrum) and *msms_spectrum_peak* (for the peak variables, i.e. the m/z and intensity values). So, I chose the name *sample_to_msms_spectrum* as database table name here to indicate that this table provides the mapping from sample to spectrum. The name *msms_spectrum* is a bit misleading indeed because that table contains just MS spectrum data (MS1, MS2, ...). The database scheme is the same that we use(d) in `CompoundDb`, and there it's mostly just MS2 spectra in that table. I just kept the same database layout so we can re-use code across packages.

Hope this explains it.

philouail replied:
so the *msms_spectrum* tables contains both MS1 and MS2 data ? So there is a column in this table that precise the msLevel ?

jorainer replied:
yes, exactly. that table contains all the data from `spectraData()` of a `Spectra` object, i.e. a column for each of the `spectraVariables()` (including `msLevel`). The database layout actually represents the structure within a `Spectra` - it's one table for the `spectraVariables` and one for the `peakVariables`. I wanted to keep it as simple as possible.

philouail replied:
Okay that clear now thanks 👍

philouail commented on R/MsExperiment-db.R:
```diff
@@ -0,0 +1,166 @@
+    if (inherits(spectra(x)@backend, "MsBackendSql")) {
+        be <- spectra(x)@backend
+        if (inherits(be, "MsBackendOfflineSql")) {
+            flags <- be@flags
```
what are the flags ?

jorainer replied:
that's specific for SQLite - it allows to have the connection either read-only or read/write. Other database systems control that type of things using a user management system (e.g. setting permissions for users for specific databases). SQLite does not provide a user management. By default, we're connecting to the databases read-only - but here we want to actually write something to the database, thus we (temporarily) set the connection to read/write. It's a bit of a hack, but with SQLite that's the only way we can do that.

philouail reviewed (approved):
For me it's completely good, and the vignette is very good. It definitely helped me when I tried it with our dataset. 

there is one part of the code that I do not understand (see comments). I'm not super familiar with the database commands yet.

jorainer committed:
docs: use recent version of MsBackendSql to simplify vignette

- With the new MsBackendSql package there is no need anymore to disable parallel
  processing in `setBackend` which simplifies usage.

jorainer committed:
ci: only run unit test on push

jorainer committed:
ci: install MsBackendSql dev version
